### ISSUE TITLE
Ranges not passed to built-in commands correctly

### ISSUE THREAD

**Issue by andmis** [linty-org/key-menu.nvim#10]
To reproduce:
- Add `require 'key-menu'.set('n', 'g')` to `init.lua`.
- Open a file with several lines.
- Press `3gg`.
- *Expected*: cursor goes to line 3.
- *Actual*: cursor goes to line 1.

The reason this happens is related to how built-ins are dealt with. Because of https://github.com/neovim/neovim/issues/18804 and https://github.com/vim/vim/issues/10503, when you press `gg`, the built-in command does not get activated directly. Instead, the `g` mapping is activated (handling the first `g` in `gg`), and then the second `g` gets sent as a keystroke to the pop-up window, which then feeds the keys `gg` back into the original buffer with no remapping, which causes the `gg` built-in to fire. We need to
1. capture any range typed by the user, and
2. forward it to the `feedkeys('gg')`.

**Comment by matu3ba** [comment]
Do you feel like https://github.com/vim/vim/issues/10503 will remain closed?

Otherwise: How should this be proceeded on? 
As I understand it, the autocommands are not usable, because there is no timeout info for in-builds.
So this requires to 1. resolve and derive the necessary info before the info is processed or 2. hacking up the vim parts for keys.
Do I miss anything here?

**Comment by andmis** [comment]
I suspect the Vim issue will remain closed. There is also a [Neovim issue](https://github.com/neovim/neovim/issues/18804) with more information – I think it's really not trivial to resolve this in core.

Right now the way that something like `ga` works is:
- We map `g` to `key_menu.open_window('g')`.
- If there is no user mapping for `ga`, then the key-menu window's `a` entry will point to the built-in `ga` command.
- If you press `a`, then the key-menu window will essentially call `feedkeys('ga')` with no remapping -- so the built in command for `ga` will be executed since no remapping happens.
- If you press `ga` quickly, it appears that the built-in `ga` command gets executed, but in fact what happens is that once you press the `a`, Vim checks for a user mapping for `ga`, doesn't find one, so it executes the `g` user mapping, which executes `key_menu.open_window`, and then immediately executes the `a` mapping, so the `a` gets passed to key-menu. I have not seen the key-menu window flashing, but it wouldn't surprise me to find out that it sometimes flashes. This is a shitty setup (IMO) but "it works".

In order to support ranges/registers with the current setup, I guess the simplest thing to try would be (1) catching them in the `open_window` step, and (2) forwarding them at the `feedkeys` step. But I haven't looked into this much. Sounds messy.

**Comment by matu3ba** [comment]
As debugging macros (by translating them into a sequence of lua function calls with arguments)
is blocked by this, I would suggest to make a plan how to fix this upstream to make neovim fully modal 
(capturing keys getting pressed and have some way to lookup the corresponding action sequence).

Any pointers whom to ask or where to look for?

**Comment by andmis** [comment]
I'm not sure exactly what it would take to do what you are asking for, but there are a lot of things messed up in the core design of the Vim mapping/command system that make it clunky to write plugins (like key-menu.nvim) dealing with mappings. I think those things tend to be non-trivial to change (https://github.com/neovim/neovim/issues/18484) and/or Bram doesn't consider them to be problems (https://github.com/vim/vim/issues/10503) and Neovim would have to diverge drastically from Vim to change/fix them.

tl;dr: good luck.
